Parameter node wrapper

Connor-Bowley opened the pull request:
Adds a Python parameter node wrapper that offers access to typed properties instead of going through strings and manually serializing/deserializing to string manually.

Adds tests for the wrapper.

Updates VectorToScalarVolume to use new parameter node as an example of use (the tests also serve as example).

I have added a `parameter_nodes.md` file documenting purpose, usage, and extensibility, but would like recommendations on where would be good to link to it from.

Closes https://github.com/Slicer/Slicer/issues/6665

Connor-Bowley commented:
Fixed linting errors.

allemangD commented on Base/Python/slicer/tests/test_slicer_parameter_node_wrapper.py:
```diff
@@ -0,0 +1,510 @@
+import vtk
+import slicer
+
+from _AverageMesh.TypedParameterNodeMk3 import (
```
Should be `from slicer.parameterNodeWraper import (`

Connor-Bowley replied:
Good catch, yes.

Connor-Bowley replied:
See https://github.com/Slicer/Slicer/pull/6684/commits/279415ebbfc582816e07a53c458918a33ff07002

allemangD commented on Base/Python/slicer/parameterNodeWrapper.py:
```diff
@@ -0,0 +1,814 @@
+    Serializer for str. This is the type that parameterNode.GetParameter/SetParameter use
+    so it is mainly a pass through.
+    """
+    @staticmethod
```
Nitpick - Making a method static in a derived class seems like a code smell to me; the signature isn't the same anymore. It seems to work, though, so it might not be an issue.

Connor-Bowley replied:
I can update to not be static to match.

Connor-Bowley replied:
See https://github.com/Slicer/Slicer/pull/6684/commits/0de798353a335df4f47c58b36e8d3eec58520dfc

allemangD commented on Base/Python/slicer/parameterNodeWrapper.py:
```diff
@@ -0,0 +1,814 @@
+        self._saveList()
+        return self
+
+    def __add__(self, other):
+        raise NotImplementedError("Adding an ObservedList is not supported. However, += is supported.")
```
Do these need to be forbidden? It seems clear to me that `data.the_list + data.the_other_list` should produce a _new_ list that is not observed.

I'm not sure how that would play with the caching mechanic, though.

I suppose `list(data.the_list) + ...` is a suitable workaround.

Connor-Bowley replied:
In my opinion, the explicit breaking of the observation helps understandability. If a consensus of people think this should be implicit, I can change it.

allemangD commented on Base/Python/slicer/parameterNodeWrapper.py:
```diff
@@ -0,0 +1,814 @@
+    def remove(self, parameterNode, name: str) -> None:
+        """Removes this parameter from the node if it exists."""
+        paramName = lambda index: self._paramName(name, index)
+        for index in range(self._len(parameterNode)):
```
Static analyzer revealed that `self._len` and `self._lenSerializer.remove` are both missing the `name` argument here.

However I'm having a hard time working the call stack to see if/how to trigger this method to see if it's actually a problem.

Connor-Bowley replied:
I will look into this. The way to trigger it should be to do a `list[list[something]]` and make the outer list smaller by setting it to smaller list

```
param.listlistx = [[1],[2]]
param.listlistx = [[3]]
```

allemangD replied:
Ah, yes it is an issue.

Suggest to add to `test_list_of_lists`:


  pl.pop(0)
  self.assertEqual(param.p, [[4, 5, 0, 7]])
  self.assertEqual(param2.p, [[4, 5, 0, 7]])

Might also consider overriding `ObservableList.__del__` to support `del pl[0]`; this is what I tried before but it does not propagate the change to the serializer.

Connor-Bowley replied:
Will add the test. Will look int overriding `__del__`.

Connor-Bowley replied:
Re: __del__

I think that `__delitem__` is the right thing here, which I implemented, but the implementation was wrong (it didn't save the updated list).

Connor-Bowley replied:
See https://github.com/Slicer/Slicer/pull/6684/commits/e3b8b690cbc93be711f1bf69fd53384b106a19eb

Connor-Bowley replied:
See also https://github.com/Slicer/Slicer/pull/6684/commits/638e6a5a0f513caccc617fa09513e522be40236f

allemangD replied:
Oops! Yes, `__delitem__` is what I meant.

allemangD reviewed (commented):
Overall looks good. I was able to play around with it.

Some nitpicks and some holdovers from previous versions that need to be addressed, but I was able to play around with it and I think it can do almost everything I'd need.

I do miss a serializer for `dict`; I think an `ObservableDict` would be a good addition at some point.

I also wonder if there's any value in a serializer for `tuple`?

Connor-Bowley commented:
I will work on an ObservableDict so we can use that, but I think it would be good to not block this PR on that (it is added functionality rather than some issue with this PR).

What are you thinking would be the need the `tuple` serialization would fill? Just non-editable lists?

allemangD commented:
> I will work on an ObservableDict so we can use that, but I think it would be good to not block this PR on that (it is added functionality rather than some issue with this PR).

I agree.

> What are you thinking would be the need the `tuple` serialization would fill? Just non-editable lists?

I wasn't thinking of anything in particular, but yes. It's a little more convenient to not have to wrap tuples returned from functions.

It would be nice for all those "fundamental" python types like `dict`, `tuple`, `set`, `bytes` etc. to be supported out-of-the-box, but the more I think about it I'm not sure the value added is worth the hassle.

I'd be comfortable merging with only these current serializers, with the expectation that more types be supported in the future if a need arises. It might make sense to convert `parameterNodeWrapper.py` into a package to try and improve organization.

Connor-Bowley commented:
> expectation that more types be supported in the future if a need arises

Definitely agree. And we can always prototype them as custom serializers where they are first needed.

> It might make sense to convert parameterNodeWrapper.py into a package to try and improve organization.

Are you suggesting like adding a `parameterNodeWrapper/` folder in `slicer` and have a setup like follows?

- `slicer/parameterNodeWrapper/__init__.py
- `slicer/parameterNodeWrapper/validators.py
- `slicer/parameterNodeWrapper/serializers.py
- etc

Or are you suggesting like a totally separate python package? I would have some reservations with that because of the strong dependency to the `vtkMRMLScriptedModuleNode` and `vtkMRMLNode`.

allemangD commented:
I meant a folder containing files like `validators.py` `listSerializer.py`, `dictSerializer.py` or similar. Maybe it would make sense to put all the base classes in some `core.py`? I don't have strong opinions on how exactly it should be organized, but I think the one big module should be broken up if we're to have even more serializers or validators in the future.

Connor-Bowley commented:
I will squash the current changes, refactor into multiple files and force push a new commit tree.

Connor-Bowley commented:
Refactored into multiple files.

Connor-Bowley commented:
One thing that is not supported is doing something like

```py
@parameterNodeWrapper
class CustomParameterNode:
  varname: list[Annotated[str, Choice(['a', 'b', 'c']]
```

which is desirable. I will investigate adding support for this.

Connor-Bowley commented:
Fixed issue where the element type for the lists couldn't be annotated. Added test for this use case.

Refactored a bit to make the addition of custom serializers more complete, giving them the ability to have their own default validators and values.

Added a new `@parameterNodeSerializer` decorator to register serializers for easy use by the parameterNodeWrapper and the things like `ListSerializer`.

Rebased off of latest main.

sjh26 commented on Docs/developer_guide/parameter_nodes.md:
```diff
@@ -0,0 +1,366 @@
+
+## Parameter Node Wrapper Classes
+
+The parameter node wrapper allow wrapping around a `vtkMRMLScriptedModuleNode` parameter node with typed member access. A simple example is as follows.
```
```suggestion
The parameter node wrapper allows wrapping around a `vtkMRMLScriptedModuleNode` parameter node with typed member access. A simple example is as follows.
```

Connor-Bowley replied:
Updated.

sjh26 commented on Docs/developer_guide/parameter_nodes.md:
```diff
@@ -0,0 +1,366 @@
+### Caching
+
+The `vtkMRMLScriptedModuleNode` stores its values as strings, but converting from string to another
+type can be slow. Caching is used to speed up repeated reads under the hood. If the value hasn't been written to the _parameter node wrapper_ then it will use the cached value. Note this means if the underlying `vtkMRMLScriptedModuleNode` parameter node changes not through the wrapper, the cache will not be updated and the cached value will be wrong.
```
```suggestion
type can be slow. Caching is used to reduce repeated costly reads under the hood. If a value hasn't been written through the _parameter node wrapper_ since the last read, then it will use the cached value. Note this means if the underlying `vtkMRMLScriptedModuleNode` parameter node changes outside of the wrapper, the cache will not be updated and the cached value will be wrong.
```

Connor-Bowley replied:
Updated.

Connor-Bowley commented:
@jcfr @lassoan @pieper Is this good to merge or did you want to review?

pieper commented on Modules/Scripted/VectorToScalarVolume/VectorToScalarVolume.py:
```diff
@@ -1,11 +1,13 @@
 import slicer
 from slicer.ScriptedLoadableModule import *
 from slicer.util import VTKObservationMixin
+from slicer.parameterNodeWrapper import *
```
I'm not a big fan of importing `*` from modules since then the symbols show up later in the code with no indication where they came from.  I'd usually prefer spelling out the full path in the code (verbose, but explicit) or if that is really too long, then listing out the explicit imports.

Connor-Bowley replied:
It kind of matched `from slicer.ScriptedLoadableModule import *` above. Could also do something like `import slicer.parameterNodeWrapper as pnw` as a shorter but still explicit middle ground.

pieper replied:
Yes, but for those cases the imported symbols are `ScriptedLoadableModuleWidget` and `VTKObservationMixin` so the full path really would be redundant.

Yes, using `pnw` would be an improvement so someone could easily track where the symbol comes from just by searching in this file.

Connor-Bowley replied:
Updated to use explicit imports.

pieper reviewed (approved):
I think this is worth getting into the preview.  It should simplify bookkeeping 👍

Connor-Bowley commented:
Rebased off of main.

jcfr commented on Modules/Scripted/VectorToScalarVolume/VectorToScalarVolume.py:
```diff
@@ -1,11 +1,17 @@
+    parameterNodeWrapper,
+    Choice,
+    Default,
+)
```
May be worth also updates the `Scripted*` templates to use the decorator. See https://github.com/Slicer/Slicer/tree/main/Utilities/Templates/Modules

To regenerate the examples in [Extensions/Testing](https://github.com/Slicer/Slicer/tree/main/Extensions/Testing), running the target `SlicerGenerateExtensionTemplates` will do it.

Connor-Bowley replied:
Will update the templates in a separate PR.

jcfr reviewed (dismissed):
Nitpick:
* documentation:
  * link documentation from "Developer Guide / MRML / Advanced topic. If helpful, table could be used if it makes the visual parsing easier.
  * using the admonition like `:::{warning}`, `:::{note}` may be helpful. Here is an example of documentation where these are used. See https://cell-locator.readthedocs.io
* Review the templates to ensure they can be updated in a follow up PR

Connor-Bowley commented:
Updated documentation to use tables. Linked to the top of the developer guide and from MRML overview advanced topics.

jcfr commented:
Thanks @Connor-Bowley for making this happen :100: :rocket: :tada:

Also thanks to @sjh26 & @allemangD, and everyone else who provided feedback and spent time reviewing :pray:

allemangD commented:
For the sake of standardization, I wonder if it would be appropriate to add some abstract classes that use this, and make the extension wizard use that abstract class. If not an abc or mixin, I think it would be good to reduce the amount of boilerplate needed to correctly use parameter nodes (preferably wrapped parameter nodes). I'd hope putting that mechanism in the extension wizard or other templates would help encourage module developers to use the same patterns and make static analysis more viable.

A rough idea of the API might be:

```python
class VectorToScalarVolume(SimpleScriptedModule):
    UI_RESOURCE_PATH = 'UI/VectorToScalarVolume.ui'
    PARAMETER_NODE_TYPE = VectorToScalarVolumeParameterNode

    def __init__(self, parent):
        super().__init__(parent)

        self.logic = None

    def setup(self):
        super().setup()  # automatically sets up `self.params`

        self.logic = VectorToScalarVolumeLogic()

        self.ui.inputSelector.currentNodeChanged.connect(self.onGUIChanged)

        ...

    def initializeParameterNode(self):  # no boilerplate; called by the base
        if not self.params.InputVolume:
            ...

    def updateGUIFromParameterNode(self):  # no boilerplate
        self.ui.inputSelector.setCurrentNode(self.params.InputVolume)
        ...

    def updateParameterNodeFromGUI(self):  # no boilerplate
        self.params.InputVolume = self.ui.inputSelector.currentNode

    def onApplyButton(self):
        ...
```

The main idea is there's no boilerplate for managing the parameter node, avoiding recursive gui/parameter updates, etc.

I'm not sure what exactly the scope of this kind of feature should be, or whether a base class is the right mechanism. Maybe a mixin, or method decorators? Either way, I don't think this kind of logic belongs in the existing `ScriptedLoadableModule` classes; some modules customize the workflow significantly and this would just get in the way.

---

I'll think on it more and wait for some preliminary feedback here... if there seems to be a reasonable way forward I'll open a new issue and try prototyping something.

Connor-Bowley commented:
I plan to update the extension wizard to use `@parameterNodeWrapper` soon.
